### PR TITLE
Fix ONNXProgram.save to use torch.load(..., mmap=True) for large models

### PR DESCRIPTION
**Pull request by thiagocrepaldi**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #117295
* #117294

During ONNXProgram.save, the implicit/explicit state_dict passed in must
be loaded in memory in order to read each initializer and create an
external tensor proto with them

This PR ensures torch.load uses memory-map to support large models that
cannot fit in memory

### CODE REVIEW

**Review by BowenBao — COMMENTED** [review]
Very nice. Do you have a case study show casing the efficiency/effectiveness before and after?

**Review by BowenBao — APPROVED** [review]

File: torch/onnx/_internal/exporter.py [hunk 1]
```diff
@@ -1002,7 +1002,10 @@ def save(
                     # ignore duplicate
                     continue
                 try:
-                    extra_state_dict = torch.load(path)
+                    # Loads checkpoint using memory-map on CPU to succeed with large models
+                    extra_state_dict = torch.load(
+                        path, map_location="cpu", mmap=True, weights_only=True
```
Comment (BowenBao) [inline torch/onnx/_internal/exporter.py:1007]: noticed `weights_only=True` is added. is this intended?

Comment (thiagocrepaldi) [reply torch/onnx/_internal/exporter.py:1007]: It is a security feature. when `weights_only=False`, malicious pickled checkpoints can execute code on the machine.
If the checkpoint can be loaded with `weights_only=True`, we should - but I am experimenting with it

**Comment by thiagocrepaldi** [comment]
> Very nice. Do you have a case study show casing the efficiency/effectiveness before and after?

Not yet.

**Pull request closed without merging**